File: orig-src/Common/Util/ObjectActivator.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using CloneExtensions;
using Fasterflect;
using QuantConnect.Securities;

package com.quantconnect.lean.Util
{
    /**
    /// Provides methods for creating new instances of objects
    */
    public static class ObjectActivator
    {
        private static final object _lock = new object();
        private static final object[] _emptyObjectArray = new object[0];
        private static final Map<Type, MethodInvoker> _cloneMethodsByType = new Map<Type, MethodInvoker>();
        private static final Map<Type, Func<object[], object>> _activatorsByType = new Map<Type, Func<object[], object>>();

        static ObjectActivator() {
            // we can reuse the symbol instance in the clone since it's immutable
            ((HashSet<Type>) CloneFactory.KnownImmutableTypes).Add(typeof (Symbol));
            ((HashSet<Type>) CloneFactory.KnownImmutableTypes).Add(typeof (SecurityIdentifier));
        }

        /**
        /// Fast Object Creator from Generic Type:
        /// Modified from http://rogeralsing.com/2008/02/28/linq-expressions-creating-objects/
        */
        /// This assumes that the type has a parameterless, default constructor
         * @param dataType">Type of the object we wish to create
        @returns Method to return an instance of object
        public static Func<object[], object> GetActivator(Type dataType) {
            lock (_lock) {
                // if we already have it, just use it
                Func<object[], object> factory;
                if( _activatorsByType.TryGetValue(dataType, out factory)) {
                    return factory;
                }

                ctor = dataType.GetConstructor(new Type[] {});

                //User has forgotten to include a parameterless constructor:
                if( ctor == null ) return null;

                paramsInfo = ctor.GetParameters();

                //create a single param of type object[]
                param = Expression.Parameter(typeof (object[]), "args");
                argsExp = new Expression[paramsInfo.Length];

                for (i = 0; i < paramsInfo.Length; i++) {
                    index = Expression.Constant(i);
                    paramType = paramsInfo[i].ParameterType;
                    paramAccessorExp = Expression.ArrayIndex(param, index);
                    paramCastExp = Expression.Convert(paramAccessorExp, paramType);
                    argsExp[i] = paramCastExp;
                }

                newExp = Expression.New(ctor, argsExp);
                lambda = Expression.Lambda(typeof (Func<object[], object>), newExp, param);
                factory = (Func<object[], object>) lambda.Compile();

                // save it for later
                _activatorsByType.Add(dataType, factory);

                return factory;
            }
        }

        /**
        /// Clones the specified instance using reflection
        */
         * @param instanceToClone">The instance to be cloned
        @returns A field/property wise, non-recursive clone of the instance
        public static object Clone(object instanceToClone) {
            type = instanceToClone.GetType();
            MethodInvoker func;
            if( _cloneMethodsByType.TryGetValue(type, out func)) {
                return func(null, instanceToClone);
            }

            // public static T GetClone<T>(this T source, CloningFlags flags)
            method = typeof (CloneFactory).GetMethods().FirstOrDefault(x -> x.Name == "GetClone" && x.GetParameters().Length == 1);
            method = method.MakeGenericMethod(type);
            func = method.DelegateForCallMethod();
            _cloneMethodsByType[type] = func;
            return func(null, instanceToClone);
        }

        /**
        /// Clones the specified instance and then casts it to T before returning
        */
        public static T Clone<T>(T instanceToClone) where T : class
        {
            clone = Clone((object)instanceToClone) as T;
            if( clone == null ) {
                throw new Exception( "Unable to clone instance of type " + instanceToClone.GetType().Name + " to " + typeof(T).Name);
            }
            return clone;
        }
    }
}

File: src/main/java/com/quantconnect/lean/util/ObjectActivator.java
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//using System.Linq.Expressions;
//using CloneExtensions;
//using Fasterflect;

package com.quantconnect.lean.util;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Supplier;

import com.quantconnect.lean.SecurityIdentifier;
import com.quantconnect.lean.Symbol;

/**
 * Provides methods for creating new instances of objects
 */
public /*static*/ class ObjectActivator {

    private static final Object[] _emptyObjectArray = new Object[0];
    private static final Map<Class<?>,Method> _cloneMethodsByType = new HashMap<>();
    private static final ConcurrentMap<Class<?>,Supplier<?>> _activatorsByType = new ConcurrentHashMap<>();
    private static final Set<Class<?>> knownImmutableTypes = new HashSet<>();
    private static final Class<?>[] NO_PARAMETER_ARRAY = new Class[0];

    static {
        // we can reuse the symbol instance in the clone since it's immutable
        knownImmutableTypes.add( Symbol.class );
        knownImmutableTypes.add( SecurityIdentifier.class );
    }

    private ObjectActivator() { }

    /**
     * Fast Object Creator from Generic Type:
     * Modified from http://rogeralsing.com/2008/02/28/linq-expressions-creating-objects/
     * This assumes that the type has a parameterless, default constructor
     * @param dataType Class of the object we wish to create
     * @returns Method to return an instance of object
     */
    @SuppressWarnings("unchecked")
    public static <T> Supplier<T> getActivator( final Class<T> dataType ) {
        // if we already have it, just use it
        return (Supplier<T>)_activatorsByType.computeIfAbsent( dataType, dt -> {
            try {
                final Constructor<T> ctor = dataType.getConstructor( NO_PARAMETER_ARRAY );

                //User has forgotten to include a parameterless constructor:
                if( ctor == null )
                    return null;

                final Supplier<T> factory = () -> {
                    try {
                        return ctor.newInstance( _emptyObjectArray );
                    }
                    catch( InstantiationException | IllegalAccessException | IllegalArgumentException
                            | InvocationTargetException e ) {
                        throw new RuntimeException( e );
                    }
                };

                return factory;
            }
            catch( NoSuchMethodException | SecurityException e ) {
                throw new RuntimeException( e );
            }
        } );
    }

    /**
     * Clones the specified instance using reflection
     * @param instanceToClone The instance to be cloned
     * @returns A field/property wise, non-recursive clone of the instance
     */
    @SuppressWarnings("unchecked")
    public static <T> T clone( final T instanceToClone ) {
        final Class<? extends Object> type = instanceToClone.getClass();
        final Method method = _cloneMethodsByType.computeIfAbsent( type, t ->
        Arrays.stream( type.getMethods() ).filter( x -> x.getName().equals( "getClone" ) && x.getParameters().length == 1 ).findFirst().orElse( null ) );

        try {
            return (T)method.invoke( null, instanceToClone ); //Static method
        }
        catch( IllegalAccessException | IllegalArgumentException | InvocationTargetException e ) {
            throw new RuntimeException( e );
        }
    }
}
